dhall/src/Dhall/ETags.hs

```diff
@@ -0,0 +1,136 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+-- | This module contains the implementation of the @dhall tags@ command
+
+module Dhall.ETags
+    ( -- * ETags
+      generate
+    ) where
+
+import Dhall.Util (Input(..))
+import Dhall.Core (Expr(..), Binding(..))
+import Dhall.Src (Src(..))
+import qualified Dhall.Map as DM (keys)
+import Dhall.Parser
+import Data.Text (Text)
+import Control.Monad
+import Control.Exception
+import qualified Data.Text as T
+import qualified Data.Text.IO as TIO
+import qualified Data.Map as M
+import Data.List (isSuffixOf)
+import System.Directory (doesDirectoryExist, getDirectoryContents,
+#if MIN_VERSION_directory(1,3,0)
+                         pathIsSymbolicLink)
+#else
+                         isSymbolicLink)
+#endif
+import System.FilePath ((</>), takeFileName)
+import Text.Megaparsec (sourceLine, sourceColumn, unPos)
```

Comment:
Can you bring the import section a bit closer to the style that the rest of the library uses? In general, qualified imports are easier to read.

Reply:
Sorry, done. :)
Do you mean, it's better to use import qualified instead of just import?
In case yes, why? Then you will have all this weird infix function like `map Data.Map.! index`.

Reply:
> Do you mean, it's better to use import qualified instead of just import?

The advantage of fully qualified imports is that you can easily understand code anywhere in a file without having to look at the imports section. This is particularly useful when reviewing diffs where the imports section is not shown.

> In case yes, why? Then you will have all this weird infix function like `map Data.Map.! index`.

That would be ok. Note that `!` by itself is already a bit of a code smell since it is partial: You'd want to have some proof that it cannot fail. If you can do without partial functions, that would be better.

Reply:
I haven't documented the style anywhere, but unqualified imports are okay as long as they are explicit, such as:

```haskell
import Data.Map ((!))
```

... in fact, I usually import operators in this way

Reply:
Done